### DISCUSSION TITLE
ES2020 in JS pb_hooks, what the best solution ?

### DISCUSSION THREAD

**Discussion by Zehir** [pocketbase/pocketbase#3352] [Q&A]
Hello, I am at a point where I dont know what direction to choose and would like some feed back on this.

My problem : I need to verify an uploaded file and insert data depending of the content. I already made a npm package to read/write the file but it's using some new functions from EC2020 like BigInt

I tried to use babel in the last 6 hours but was not able to make it work to bundle the hook with the external package.

So there is what I can do :
1) persist with babel to compile to ES5, no success with that, may never work
2) build a small internal webserver using Bun to convert the file to json and use the json from the hook, probably easiest but not performant ? Probably not an issues as they are really light to process  The thing I dont like with that is to forward the uploaded file again to an other webserver to parse it. Seems slow but maybe not
3) compile the npm package in go and use go hooks, but need to learn go...
4) hope for ES2020 support in pocket base

I searched on internet but was not able to find any example with rollup for pocket base, if anybody made this I would like to see some

Thanks

**Comment by ganigeorgiev** [comment]
>  I already made a npm package to read/write the file but it's using some new functions from EC2020 like BigInt
> I tried to use babel in the last 6 hours but was not able to make it work to bundle the hook with the external package.

I'm not entirely sure that I understand what do you mean by this. Could you provide a code sample?


In general modifying/intercepting uploaded files is not very user friendly at the moment.

For an example with Go you can check https://github.com/pocketbase/pocketbase/discussions/3315.

For the JS `pb_hooks` this is even more difficult (_and maybe even impossible depending on what you want to do with the file due to the lack of exposed helpers around the Go reader/writer interfaces; I'll consider exposing some of the `io` package functions as JSVM bindings with the next v0.19/v0.20 releases_).

**Comment by Zehir** [comment]
The goal here is to store configuration files for an app (deconz), we call them bundle.
The bundle format is described here : https://github.com/deconz-community/ddf-tools/tree/main/packages/bundler
If you want to make one you can use this page : https://deconz-community.github.io/ddf-tools/bundler
- Click on one of the button on the `Build from GitHub source` and then press `Build`

I can't really trust the user about what inside the bundle so I need to extract it, validate it's content and then create the records in the database in multiple collections. The problem is that the bundle is a binary files so I need to decode it and there is also some signatures process.

So my plan is not to intercept the file on collection update but directly a new api endpoint to upload bundle and that endpoint handle the database part.

I am using the packages `@noble/curves` and `pako` and one of them is using the [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) method that was added in [ES2020](https://www.w3schools.com/js/js_2020.asp)
So when building the hooks, from inside pocketbase `BigInt` is undefined and the bundler don't work.

You can see my current code of the hooks here :
https://github.com/deconz-community/ddf-tools/tree/main/packages/store
The pb_hooks directory is created when compiling with rollup

**Reply by ganigeorgiev** [reply]
I'm not sure that I'll be able to help you with the rollup/babel config, but just by looking at some of the `@noble/curves`, `@noble/hash` and `paco` code, it is not only `BigInt` that you'll have to polyfill (they also seems to rely on a lot of `crypto` primitives that even a polyfill may not be enough) and in my opinion you may find it easier switching to Go.

**Reply by Zehir** [reply]
Can I make go hooks and js hooks at the same time ?
The issue here is that I don't know go..

**Reply by Zehir** [reply]
And can I expose go methods to js hooks ?

**Reply by ganigeorgiev** [reply]
> Can I make go hooks and js hooks at the same time ?

Yes, the JS engine integration is implemented as "plugin" and can be enabled programmatically. For an example see the main file of the prebuilt executable: https://github.com/pocketbase/pocketbase/blob/f605521208179072a1251042e2b16835fcc955cc/examples/base/main.go#L95-L101

---

> And can I expose go methods to js hooks ?

No, this is not possible at the moment.

**Reply by Zehir** [reply]
~~Did you think about using :~~
~~https://github.com/openwebf/quickjs~~
~~Instead of~~
~~https://github.com/dop251/goja_nodejs~~

~~The original project of quickjs has been restarted by openwebf and he support ES2020 :)~~
~~I notice that you don't have time for the next month to work actively on pocketbase. And I don't know what kind of work is needed to clone the plugin and make my own using quickjs~~

Nevermind I missread, quickjs is not a go module

Well maybe https://github.com/quickjs-go/quickjs-go

**Comment by Zehir** [comment]
For now my best way to do it will be to use some `$os.exec` to send and get back the result
